### PR TITLE
[ML] relax throttling on expired data cleanup

### PR DESCRIPTION
**Pull request by benwtrent**
Throttling nightly cleanup as much as we do has been over cautious.

Night cleanup should be more lenient in its throttling. We still
keep the same batch size, but now the requests per second scale
with the number of data nodes. If we have more than 5 data nodes,
we don't throttle at all.

These numbers do seem magical...maybe it is better to not throttle
at all...

### CODE REVIEW

**Comment by benwtrent** [comment]
@droberts195 it almost seems like we should add a setting for this. I am not too keen on these "magic" values.

**Comment by hendrikmuhs** [comment]
> These numbers do seem magical...maybe it is better to not throttle
> at all...

FYI

This could be a usecase for #37867 in some way, e.g. if the query part of the DBQ is running with lower priority, it will run slower and therefore also delete slower. This could supersede the  idea of throttling.

https://github.com/elastic/elasticsearch/issues/37867#issuecomment-628071342 contains an interesting answer about "Different thread pool for system indices", which touches our case, however we want lower priority, not higher. Should we add our usecase there? I leave that up to @droberts195.

**Comment by droberts195** [comment]
> it almost seems like we should add a setting for this. I am not too keen on these "magic" values.

Yes, good idea.  We need some configurability.

I think it's best if the `DeleteExpiredDataAction` takes two request arguments:

1. Maximum runtime, default 8 hours
2. Requests per second for results DBQ, default null, i.e. no limit

Then move the logic for setting the requests per second for results DBQ in nightly maintenance to the nightly maintenance code, which supplies it as a request argument when it calls `DeleteExpiredDataAction`.

This will mean that whatever our nightly maintenance does by default, a user can catch up without throttling if required by directly calling delete expired data themselves.

We could also have a dynamic cluster-wide setting to control the requests per second for results DBQ when called from nightly maintenance.  The default could be -1, meaning use the magic value, and any other value would get passed to `DeleteExpiredDataAction` by the nightly maintenance code.

**Comment by droberts195** [comment]
> This could be a usecase for #37867 in some way, e.g. if the query part of the DBQ is running with lower priority, it will run slower and therefore also delete slower. This could supersede the idea of throttling.

Yes, true, if we could run the DBQ searches with lower priority then that could be an alternative to throttling in the future.  But we need to do _something_ now, as people are running into the problem of us over throttling in big deployments.
> https://github.com/elastic/elasticsearch/issues/37867#issuecomment-628071342 contains an interesting answer about "Different thread pool for system indices", which touches our case, however we want lower priority, not higher. Should we add our usecase there?

Our state and results indices will be hidden indices, not system indices.  Also, I imagine a lot of the burden of deleting millions of documents is in disk access and Lucene segment management.  So I am not sure we should complicate that proposal with our cleanup requirements.

**Comment by davidkyle** [comment]
> I think it's best if the DeleteExpiredDataAction takes two request arguments:

Consider `job_id` as a third option as it would effectively slice the work up and may help users get over the hump

**Comment by droberts195** [comment]
> Consider `job_id` as a third option as it would effectively slice the work up and may help users get over the hump

Good idea.  It would have to be wildcardable, with a default of `*`, but it's true that it would be great for the case where one particular job is huge and either the user needs to given that extra attention or alternatively just clean up the other jobs with little effort while some other (higher effort) mechanism is used on the huge job.

**Comment by davidkyle** [comment]
I'll raise a separate PR for the `job_id` change

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/retention/UnusedStateRemover.java
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/retention/UnusedStateRemover.java:106]: Good 👀 I stared at this for a long time before I saw the difference

**Review by davidkyle — APPROVED** [review]
LGTM

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/DeleteExpiredDataRequest.java
Comment (benwtrent) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/ml/DeleteExpiredDataRequest.java:53]: GRAMMAR! My old nemesis.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlInitializationService.java
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlInitializationService.java:102]: 🤔 good point. The setting updater might have to be in the initialization service itself...

To my knowledge there is no way to deregister a setting consumer. The consumers have no unique identification.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/DeleteExpiredDataAction.java
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/DeleteExpiredDataAction.java:105]: you are correct. This is left over from when I had `requestsPerSecond` an unboxed value :).

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlInitializationService.java
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlInitializationService.java:102]: Why is this being set to `null` anyways? I assume so it can be GC'd, but the object itself is not HUGE, and only really has references to things are referenced by the `MlInitializationService.java`. 

I am gonna change the code so that it does not get set `null`. It seems like a waste to me. 

@dimitris-athanasiou @droberts195 ^ Let me know if you have a prevailing opinion the other way.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/DeleteExpiredDataAction.java
Comment (droberts195) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/DeleteExpiredDataAction.java:42]: We would usually error on unknown fields when parsing REST requests.  Should this be `false` on the server side?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/DeleteExpiredDataRequest.java
Comment (droberts195) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/ml/DeleteExpiredDataRequest.java:53]: It looks like the magic value is `-1.0f` in the core code, not `1.0f`.  Negative also makes more sense for the magic value.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteExpiredDataAction.java
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteExpiredDataAction.java:78]: Maybe rename `MAX_DURATION` to `DEFAULT_MAX_DURATION` now it's not always used.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/MLRequestConvertersTests.java
Comment (droberts195) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/MLRequestConvertersTests.java:218]: Might be better to test with values other than 1, as 1 is more likely to end up in the output due to the combination of a bug and a fluke.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/DeleteExpiredDataRequest.java
Comment (davidkyle) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/ml/DeleteExpiredDataRequest.java:53]: Good point. I think null (unspecified) means use the default and in the action that is interpreted as the magic value -1.0f

Comment (benwtrent) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/ml/DeleteExpiredDataRequest.java:53]: `null` means `no throttle`

`-1.0f` is our "magic" calculation.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/DeleteExpiredDataAction.java
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/DeleteExpiredDataAction.java:42]: yes yes, it should be false. Fixing!

**Comment by benwtrent** [comment]
@elasticmachine update branch

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/action/TransportDeleteExpiredDataActionTests.java
Comment (droberts195) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/action/TransportDeleteExpiredDataActionTests.java:100]: It seems dodgy to test two error conditions together: a timeout and requests per second = 0.  Maybe the `.0f` was a typo?  I think this test should just test the timeout.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/DeleteExpiredDataRequest.java
Comment (droberts195) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/ml/DeleteExpiredDataRequest.java:53]: "default" is probably the wrong word for this comment.  The other HLRC docs say the default is `null`, and that's what's implemented in the HLRC.

So maybe "default" to "standard nightly maintenance behavior" or something like that.

**Review by droberts195 — APPROVED** [review]
LGTM

Thanks for doing this impromptu piece of work, but I think it's something a few users need urgently.

I saw a couple of nits but am happy for you to merge this without another review.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/DeleteExpiredDataRequest.java
Comment (benwtrent) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/ml/DeleteExpiredDataRequest.java:53]: Definitely, saying `default` is a bit disingenuous.

**Pull request merged**
